GWXtreme/eos_prior.py

```diff
@@ -0,0 +1,338 @@
+    import lalsimulation
+
+    
+    gamma1, gamma2, gamma3, gamma4 = eos_parameters
+    eos = lalsimulation.SimNeutronStarEOS4ParameterSpectralDecomposition(
+        gamma1, gamma2, gamma3, gamma4,
+    )
+    
+
+    return eos
+
+def polytrope_eos(eos_parameters: tuple) -> typing.Any:
+    import lalsimulation
+
+    logP1, gamma1, gamma2, gamma3 = eos_parameters
+
+    return lalsimulation.SimNeutronStarEOS4ParameterPiecewisePolytrope(
+        logP1,
+        gamma1, gamma2, gamma3,
+    )
+
+
+def spectral_eos_adiabatic_index(x,spectral_parameters):
+    x_sq = x * x
+    x_cu = x_sq * x
+
+    gamma1, gamma2, gamma3, gamma4 = spectral_parameters
+
+    log_gamma = (
+        gamma1 +
+        gamma2 * x +
+        gamma3 * x_sq +
+        gamma4 * x_cu
+    )
+
+    return numpy.exp(log_gamma)
+
+
+_x_min = 0.0
+_x_max = 12.3081
+_x_grid = numpy.linspace(_x_min, _x_max, 500)
+
+
+
+def is_valid_adiabatic_index(spectral_parameters: tuple):
+    # Confirm that the adiabatic index is within a tolerable range.
+    Gamma = spectral_eos_adiabatic_index(_x_grid, spectral_parameters)
+    return (0.6 < Gamma).all() and (Gamma < 4.5).all()
+
+
+
+
+def has_enough_points(eos: typing.Any) -> bool:
+    import lalsimulation
+
+    min_points = 8
+
+    logpmin = 75.5
+    logpmax = numpy.log(lalsimulation.SimNeutronStarEOSMaxPressure(eos))
+
+    dlogp = (logpmax - logpmin) / 100
+
+    m_prev = 0.0
+
+    for i in range(min_points):
+        p = numpy.exp(logpmin + i*dlogp)
+
+        r, m, k = lalsimulation.SimNeutronStarTOVODEIntegrate(p, eos)
+
+        if m <= m_prev:
+            return False
+
+        m_prev = m
+
+    return True
+
+def eos_max_sound_speed(eos: typing.Any, eos_fam: typing.Any) -> float:
+    import lalsimulation
+
+    # Maximum allowed mass
+    m_max_kg = lalsimulation.SimNeutronStarMaximumMass(eos_fam)
+    # Central pressure
+    p_max = lalsimulation.SimNeutronStarCentralPressure(m_max_kg, eos_fam)
+    # Pseudo-enthalpy at the core.
+    h_max = lalsimulation.SimNeutronStarEOSPseudoEnthalpyOfPressure(p_max, eos)
+    # Maximum sound speed should occur at the max pseudo-enthalpy for a typical
+    # TOV neutron star.
+    return lalsimulation.SimNeutronStarEOSSpeedOfSoundGeometerized(h_max, eos)
+
+
+def is_causal_eos(eos: typing.Any, eos_fam: typing.Any) -> bool:
+    c_max = eos_max_sound_speed(eos, eos_fam)
+
+    # Confirm that the sound speed is less than speed of light, with an added
+    # buffer (10%) to account for imperfect modeling.
+    c_buffer = 0.1
+    return c_max < (1.0 + c_buffer)
+
+
+def eos_mass_range(eos_fam: typing.Any) -> typing.Tuple[float,float]:
+    import lal
+    import lalsimulation
+
+    m_min = lalsimulation.SimNeutronStarFamMinimumMass(eos_fam)
+    m_max = lalsimulation.SimNeutronStarMaximumMass(eos_fam)
+
+    return m_min / lal.MSUN_SI, m_max / lal.MSUN_SI
+
+
+
+#########################
+#  Combined EoS Priors  #
+#########################
+
+
+
+
+def is_valid_eos(
```

Comment:
Please write a detailed docstring for this function, since this does the main heavy lifting. Please use the structure we have been using in the `eos_model_selection.py` code when adding the docstring.